### ISSUE TITLE
Floating Point overflow in Cell Key calculation gives results which vary with order of microdata

### ISSUE THREAD

**Issue by duncanharris** [sdcTools/UserSupport#145]
This is a bug I spotted from perusing the source code at https://github.com/ppdewolf/libtauargus

When you add the Cell Keys, e.g.:
https://github.com/ppdewolf/libtauargus/blob/15ccf3ce60d51dda0498dec9aba92b533a0df3cb/DataCell.h#L91

You will eventually overflow your `double` and lose significant digits from the rowkeys that you are adding together: you get rounding.

The consequence of this is that the cell key that you end up with can change when you keep the same row keys but only change the order in which you add them up.

A different cell key can obviously mean different perturbation. Hence the output table can change when you only change the order of the microdata lines and not any of the contents of the lines. The output table should be independent of the order of the microdata lines.

**Comment by ppdewolf** [comment]
Just to show my minimal knowledge about this: as far as I understand a floating point overflow occurs when the value exceeds the maximum possible floating point value, which is approximately 3.4E38 for single precision 32 bit. Since our record keys are values between 0 and 1, I guess we need several times the global population contributions to have an overflow. Or am I missing the point? 😉 

Moreover, the same would hold for any variable that is added for a table cell (e.g. response, shadow, cost).

**Comment by ppdewolf** [comment]
By the way: the fact that we are using record keys in [0, 1] differs from the original idea for implementation by ABS.

**Comment by duncanharris** [comment]
Actually it is worse than I thought due to the floating point representation. The difference in totalling shows with even a small quantity of numbers added together.

My C++ is fairly rusty, but try this: https://rextester.com/AYHMFF30205

Which does:
* generates 10 random doubles between 0 and 1
* for each one:
  * render to a string at 15 decimal places
  * print out the string so you can see it
  * parse that string using `atof` (to simulate what Tau Argus does)
* then sum the 10 parsed doubles and print the result
* shuffle the list and sum again (repeat x 10)

You will see that you typically get some different values for the sum even for only 10 input numbers.

I think you need to stop using floating point arithmetic. Either use integers and do the cell key modulus after each addition or use some arbitrary precision library, e.g. https://en.wikipedia.org/wiki/List_of_arbitrary-precision_arithmetic_software. I think integers are preferable. Much easier to understand and the code will run faster.

Code:
```C++
#include <iostream>
#include <iomanip>
#include <string>
#include <sstream> 
#include <vector>
#include <random>
#include <numeric>
#include <algorithm>
#include <stdlib.h>

int main()
{
    std::random_device rd;
    std::mt19937 e2(rd());
    std::uniform_real_distribution<double> dist(0, 1);
    std::vector<double> rvs;
    
    std::cout << "Random numbers:" << std::endl << std::endl;
    for (int i = 0; i < 10; i++) {
        std::stringstream ss;
        ss << std::fixed << std::setprecision(15) << dist(e2);
        std::string rv_str = ss.str();
        std::cout << rv_str << std::endl;
        double parsed_rv = atof(rv_str.c_str());
        rvs.push_back(parsed_rv);
    }
    std::cout << std::endl << "Sums:" << std::endl << std::endl;
    
    for (int i = 0; i < 10; i++) {
        double total = std::accumulate(rvs.begin(), rvs.end(), decltype(rvs)::value_type(0));
        std::cout << std::fixed << std::setprecision(15) << total << std::endl;

        std::random_shuffle(rvs.begin(), rvs.end());
    }
}
```

Sample output:
```
Random numbers:

0.641411669152032
0.095798393901968
0.941414828506820
0.820465044299236
0.272846756816880
0.342734266289129
0.224285901484730
0.309883002748663
0.968207916744879
0.219018926834018

Sums:

4.836066706778354
4.836066706778356
4.836066706778355
4.836066706778356
4.836066706778355
4.836066706778355
4.836066706778355
4.836066706778355
4.836066706778356
4.836066706778356
```

**Comment by ppdewolf** [comment]
Thanks for the example. I agree this looks scary. 

First of all, let me say that I am not a professional programmer. I'm a methodologist that likes to do programming and has gained some experience over the last 20 years or so 😉 

Since our cellkey in the end will be the fractional part of the sum of the record keys, another way to limit the rounding error is to keep the fractional part only when adding two record keys. This still gives a rounding error, but when adding 1,000,000 random numbers this way, the resulting fractional part differs only in the last 3 digits in a 15 digit representation:

```C
double myfunction(double x, double y){
        double dum;
        return modf(x+y, &dum);
}

int main()
{
    std::random_device rd;
    std::mt19937 e2(rd());
    std::uniform_real_distribution<double> dist(0, 1);
    std::vector<double> rvs;
    
    std::cout << "Random numbers:" << std::endl << std::endl;
    for (int i = 0; i < 1000000; i++) {
        std::stringstream ss;
        ss << std::fixed << std::setprecision(15) << dist(e2);
        std::string rv_str = ss.str();
        //std::cout << rv_str << std::endl;
        double parsed_rv = atof(rv_str.c_str());
        rvs.push_back(parsed_rv);
    }
    std::cout << std::endl << "Sums:" << std::endl << std::endl;
    
    for (int i = 0; i < 10; i++) {
        double total = std::accumulate(rvs.begin(), rvs.end(), decltype(rvs)::value_type(0),myfunction);
        std::cout << std::fixed << std::setprecision(15) << total << std::endl;

        std::random_shuffle(rvs.begin(), rvs.end());
    }
}
```
gives
```
Sums:

0.170852552773234
0.170852552773070
0.170852552773051
0.170852552773148
0.170852552773134
0.170852552773064
0.170852552773146
0.170852552773132
0.170852552773082
0.170852552773088
```
Or even better: round to 15 decimals AND take the fractional part only after each addition; so use
```C
double myfunction(double x, double y){
        double pwr = pow(10, 15);
        double dum;
        return modf(floor((x+y) * pwr + 0.5) / pwr, &dum);
}
```
to get (again with 1,000,000 random numbers)
```
Sums:

0.737166540279109
0.737166540279109
0.737166540279109
0.737166540279109
0.737166540279109
0.737166540279109
0.737166540279109
0.737166540279109
0.737166540279109
0.737166540279109
```

I think the easiest way out for the short term, is to adjust the summation of record keys like above and to suggest the user to supply random numbers with at most 10 digits (what would be appropriate for most instances) and also p-table values with at most 10 digits.
For the long term we might think about integer arithmetic, but that would either require differnt input from the users, or additional programming from our side to translate Uniform(0,1) variables with more than 9 digits to integers... (32 bit integer can't deal with that, although I intend to move to 64 bit in the near future.)

P.S. I was a bit confused by the "floating point overflow". As far as I know the described behaviour is due to the limited precision of doubles and the fact that numbers cannot always be represented by a finite number of (binary) digits. I always thought that "Floating point overflow" means that the number exceeds the maximum possible floating point number.

**Comment by duncanharris** [comment]
I guess the correct term is "round-off error" rather than overflow: https://en.wikipedia.org/wiki/Round-off_error

I do agree that the strategy you have reached of: "add, round, mod" for each row key does seem to work fine and will do for now.

Integers at 64 bits would be faster of course and relatively easy to do:
- take fraction after decimal point in row key and make it 15 digits long either by padding with zeros or truncating
- parse these 15 digits as an integer
- mod by 1E15 after each addition

**Comment by ppdewolf** [comment]
Thanks for the hint for 64 bits situation. Will put this on the "to-do" list when moving to 64 bit version.
I close this issue for now.

**Comment by duncanharris** [comment]
Surely you should keep this issue open until you have released a version which has fixed the bug, i.e. implemented your "add, round, mod" scheme?

**Comment by ppdewolf** [comment]
Is fixed in sdcTools/libtauargus@b1da90c147ce7b55ae23d0968cf4fbdf4bd9b397 and will be in next version.
